{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3ca592bc-0854-4434-af8a-289631fcc8fa",
   "metadata": {},
   "source": [
    "# Parsing Genbank annotations\n",
    "## Extracting gene boundaries for variant annotation\n",
    "### August 13th, 2025\n",
    "\n",
    "This notebook aims to explore parsing Genbank records for variant annotations. Usually a Genbank record represents a single segment with (among other things) annotations of coding regions. There can be multiple genes in one segment, and nucleotides in genes need not be contiguous - splicing can occur, as well as ribosomal slippage sites. These result in the need to correctly parse `CDS` regions, with or without `joins` such as `join(26..51,740..1007)`. This can all be handled by the Biopython library. This work was motivated by the Philadelphia Canine H3N2 sequencing, so we will therefore be exploring a run from the data as of the date of this notebook, as well the `canineh3n2` reference. We will look at the following to ensure comprehensiveness:\n",
    "\n",
    "- HA ([OR738644.1](https://www.ncbi.nlm.nih.gov/nuccore/OR738644.1/)), single, contiguous gene\n",
    "- PA ([OR738643.1](https://www.ncbi.nlm.nih.gov/nuccore/OR738643.1/)), gene with a ribosomal slippage site\n",
    "- PB1 ([OR738642.1](https://www.ncbi.nlm.nih.gov/nuccore/OR738642.1/)), multiple genes on one segment\n",
    "- MP ([OR738647.1](https://www.ncbi.nlm.nih.gov/nuccore/OR738647.1/)), contains spliced genes\n",
    "\n",
    "The following function is meant to be a prototype of a \"one size\" fits all parser for the scenarios described above. The return format is list of dictionaries with the following keys:\n",
    "- `name` comes from the gene qualifier if available, otherwise \"unknown\".\n",
    "- `coding` is the joined nucleotide string (strand-aware).\n",
    "- `positions` is a list of 0-based genomic positions in the original segment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2fdcd2d-d77d-4ffe-a42c-c0eaf05adbc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import CompoundLocation\n",
    "\n",
    "\n",
    "def extract_cds_with_coords(genbank_file):\n",
    "    cds_data = []\n",
    "    record = SeqIO.read(genbank_file, \"genbank\")\n",
    "\n",
    "    for feature in record.features:\n",
    "        if feature.type != \"CDS\":\n",
    "            continue\n",
    "\n",
    "        seq_parts = []\n",
    "        coord_parts = []\n",
    "\n",
    "        # Handle joined locations vs single intervals\n",
    "        if isinstance(feature.location, CompoundLocation):\n",
    "            parts = feature.location.parts\n",
    "        else:\n",
    "            parts = [feature.location]\n",
    "\n",
    "        for part in parts:\n",
    "            start = int(part.start)\n",
    "            end = int(part.end)\n",
    "            strand = part.strand\n",
    "\n",
    "            if strand == 1:\n",
    "                seq_parts.append(str(record.seq[start:end]))\n",
    "                coord_parts.extend(range(start, end))\n",
    "            else:\n",
    "                seq_parts.append(str(record.seq[start:end].reverse_complement()))\n",
    "                coord_parts.extend(reversed(range(start, end)))\n",
    "\n",
    "        coding_seq = \"\".join(seq_parts)\n",
    "        assert len(coding_seq) == len(coord_parts)\n",
    "\n",
    "        gene_name = (\n",
    "            feature.qualifiers.get(\"gene\", [record.id])[0]\n",
    "            if \"gene\" in feature.qualifiers\n",
    "            else record.id\n",
    "        )\n",
    "\n",
    "        cds_data.append({\n",
    "            \"gene\": gene_name,\n",
    "            \"coding\": coding_seq,\n",
    "            \"positions\": coord_parts\n",
    "        })\n",
    "\n",
    "    return cds_data\n",
    "\n",
    "\n",
    "def print_translations(segment):\n",
    "    gb_filepath = '../data/reference/%s/metadata.gb' % segment\n",
    "    gb = extract_cds_with_coords(gb_filepath)\n",
    "    for gene in gb:\n",
    "        print(gene['gene'], 'gene -', Seq(gene['coding']).translate())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1082248-d7e6-4068-81ad-589154cf8447",
   "metadata": {},
   "source": [
    "HA compares favorably with the translation in the Genbank record..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88d1e83f-f914-4f4f-8a27-3922ac2d8828",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HA gene - MKIVIALSYIFCLAFSQNLLGNENNAATLCLGHHAVPNGTMVKTITDDQIEVTNATELVQNSSTGKICNNPHKILDGRDCTLIDALLGDPHCDIFQNETWDLFVERSNAFSNCYPYDVPDYASLRSIIASSGTLEFITEGFTWAGVTQNGGSGACKRGPANSFFSRLNWLTKSGNTYPVLNVTMPNTNNFDKLYIWGVHHPSTDQEQTSLYIQASGRVTVSTRRSQQTIIPNIGSRPLVRGQSGRISVYWTIVKPGDILVINSNGNLIAPRGYFKMHIGKSSIMRSDAPIDTCISECITPNGSIPNEKPFQNVNKITYGACPKYVKQNTLKLATGMRNVPERQTRGLFGAIAGFIENGWEGMVDGWYGFRHQNSEGTGQAADLKSTQAAIDQINGKLNRVIEKTNEKFHQIEKEFSEVEGRIQDLEKYVEDTKIDLWSYNAELLVALENQNTIDLTDSEMNKLFEKTRRQLRENAEDMGNGCFKIYHKCDNACIESIRNGTYDHNIYRDEAVNNRFQIRGVELKSGYKDWILWISFAISCFLLCVVLLGFILWACQRGNIRCNICI*\n"
     ]
    }
   ],
   "source": [
    "print_translations('ha')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "138536a7-74ec-4859-a046-7f1578f07f54",
   "metadata": {},
   "source": [
    "As does PA for the main gene and the frameshift variant..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9bb08851-e654-4693-8296-10579797fa0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PA gene - MEDFVRQCFNPMIVELAEKAMKEYGEDPKIETNKFAAICTHLEVCFMYSDFHFIDERGESIIVEHGDPNALLKHRFEIIEGRDRTMAWTVVNSICNTTEVEKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSEKTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMASRGLWDSFRQSERGEETIEERFEIAGTMRRLADQSLPPNFSSLENFRAYVNGFEPNGYIDGKLSQMSKEVNARIEPFLKTTPRPLRLPDGPPCTQRSKFLLMDALKLSIEDPSHEGEGIPLYDAIKCMKTFFGWKEPNIIKPHKKGINPNYLLAWKQVLAELQDIENEEKIPKTKNMKKTSQLKWVLGENMAPEKVDFEDCKDVGDLKQYDSDEPKPKSLASWIQSEFNKACELTDSSWIELDEIGEDIAPIEHIASKRRNYFTAEVSHCRATEYIMKGVYINTALLNASCAAMDDFQLIPMISKCRTKEGRRKTNLYGFIIKGRSHLRNDTDVVNFVSMEFSLTDPRLEPHKWEKYCVLEIGDMLLRTAIGQVSRPMFLYVRTNGTSKIKMKWGMEMRRCLLQSLQQIESMIEAESSVKEKDMTKEFFENRSETWPIGESPKGVEEGSIGKVCRTLLAKSVFNSLYASPQLEGFSAESRKLLLIVQALRDNLEPGTFDFEGLYEAIEECLINDPWVLLNASWFNSFLTHALK*\n",
      "PA-X gene - MEDFVRQCFNPMIVELAEKAMKEYGEDPKIETNKFAAICTHLEVCFMYSDFHFIDERGESIIVEHGDPNALLKHRFEIIEGRDRTMAWTVVNSICNTTEVEKPKFLPDLYDYKENRFIEIGVTRREVHIYYLEKANKIKSEKTHIHIFSFTGEEMATKADYTLDEESRARIKTRLFTIRQEMASRGLWDSFVSPKEAKRQLKKGLKSQGPCAGLPTKVSHRTSPALKILEPM*\n"
     ]
    }
   ],
   "source": [
    "print_translations('pa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "407cd74f-d905-4233-9fe2-9cee180738f4",
   "metadata": {},
   "source": [
    "As do both genes on PB1..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a2561fb9-8af0-43c9-8483-59f1a4418ac5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PB1 gene - MDVNPTLLFLRVPAQNAISTTFPYTGDPPYSHGTGTGYTMDTVNRTHQYSEKGKWTTNTETGAPQLNPIDGPLPEDNEPSGYAQTDCVLEAMAFLEKSHPGIFENSCIETMEVVQQTRVDKLTQGRQTYDWTLNRNQPAATALANTIEVFRSNGLTANESGRLIDFLKDVMESMDKEEMEITTHFQKKRRVRDNMTKKMITQRTIGKKKQRLNKKNYIIRALTLNTMTKDAERGKLKRRAIATPGMQIRGFVYFVETLARSICEKLEQSGLPVGGNEKKAKLANVVRKMMTNSQDTELSFTITGDNTKWNENQNPRMFLAMITYITRNQPEWFRNVLSIAPIMFSNKMARLGKGYMFESKNMKLRTQIPAEMLANINLKYFNESTRKKIEKIRPLLTDGTASLSPGMMMGMFNMLSTVLGVSILNLGQKRYTKSTYWWDGLQSSDDFALIVNAPNHEGIQAGVDRFYRICKLVGINMSKKKSYINRTGTFEFTSFFYRYGFVANFSMELPSFGVSGVNESADMSIGVTVIKNNMINNDLGPATAQMALQLFIKDYRYTYRCHRGDTQIQTRRSFELKKLWEQTRSKAGLLVSDGGPNLYNIRNLHIPEVCLKWELMDEDYQGRLCNPLNPFVSHKEIESVNNAVVMPAHGPAKSMEYDAVTTTHSWIPKRNRSILNTSQRGILEDEQMYQKCCNLFEKFFPSSSYRRPVGISSMVEAMVSRAQIDARIDFESGRIKKEEFAEIMKICSTIEELRRQK*\n",
      "PB1-F2 gene - MEQGQDTPWTQSIEHINTQRKENGQQTQKLEHPNSIQLMDHYPRIMSQADMHKQIVYWKQWLSLKSPTQGSLKTHVLKRWKLFSKQEWTN*\n"
     ]
    }
   ],
   "source": [
    "print_translations('pb1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bbb3ea1-2e7c-4fe0-9179-cc0f63e80b29",
   "metadata": {},
   "source": [
    "As do both genes on MP!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42103c2a-9239-40e3-bcce-13e60e56b251",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M2 gene - MSLLTEVETPTKNEWECKCSDLSDPLIIAASIIGILHLILWILDRLFFKCIYRHLKYGLKRGPSTEGVPESMREEYRQEQQSAVDVDDGHFVNIELE*\n",
      "M1 gene - MSLLTEVETYVLSIIPSGPLKAEIAQRLEDVFAGKNTDLEALMEWLKTRPILSPLTKGILGFVFTLTVPSERGLQRRRFVQNALNGNGDPNNMDKAVKLYRKLKKEITFHGAKEVALSYSTGALASCMGLIYNRMGTVTTEVAFGLVCATCEQIADSQHRSHRQMVTTTNPLIRHENRMVLASTTAKAMEQMAGSSEQAAEAMEVANQARQMVQAMRTIGTHPSSSAGLKDDLLENLQAYQKRMGVQMQRFK*\n"
     ]
    }
   ],
   "source": [
    "print_translations('mp')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e97db9be-5e08-45b1-a8ab-8f93fdd9f76d",
   "metadata": {},
   "source": [
    "These match the respective translations reported in the Genbank files. We now use this to do a refactor of the `extract_coding_regions` and `annotate_amino_acid_changes` functions.\n",
    "\n",
    "`extract_coding_regions` will now take in a list of all Genbank files, keyed by segment, and return a dictionary, keyed first by segment, then by gene. It is essentially an aggregation of `extract_cds_with_coords` across segments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4977682e-6b94-487d-a09d-639a8908fcef",
   "metadata": {},
   "outputs": [],
   "source": [
    "segments = [\"ha\", \"mp\", \"na\", \"np\", \"ns\", \"pa\", \"pb1\", \"pb2\"]\n",
    "\n",
    "import json\n",
    "\n",
    "\n",
    "def extract_coding_regions(segments):\n",
    "    coding_regions = {}\n",
    "    for segment in segments:\n",
    "        genbank_filepath = '../data/reference/%s/metadata.gb' % segment\n",
    "        coding_regions[segment] = extract_cds_with_coords(genbank_filepath)\n",
    "    return coding_regions\n",
    "\n",
    "\n",
    "with open('coding_regions.json', 'w') as f:\n",
    "    coding_regions = extract_coding_regions(segments)\n",
    "    json.dump(coding_regions, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf956b0b-aada-459d-8f4b-71be90ed47b0",
   "metadata": {},
   "source": [
    "We now refactor `annotate_amino_acid_changes` to use this new coding regions format. Note that VarScan reports variants in 1-based coordinates, so we must offset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9db8a710-0c84-438b-9ff5-f3b50123904a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "import re\n",
    "\n",
    "\n",
    "def lookup_site(arr, site):\n",
    "    # return the location of a site in an array, or None if it is not present\n",
    "    return next(\n",
    "        (i for i, val in enumerate(arr) if val == site),\n",
    "        None\n",
    "    )\n",
    "\n",
    "\n",
    "amino_acid_abbreviations = {\n",
    "    \"A\": \"Ala\",\n",
    "    \"R\": \"Arg\",\n",
    "    \"N\": \"Asn\",\n",
    "    \"D\": \"Asp\",\n",
    "    \"C\": \"Cys\",\n",
    "    \"Q\": \"Gln\",\n",
    "    \"E\": \"Glu\",\n",
    "    \"G\": \"Gly\",\n",
    "    \"H\": \"His\",\n",
    "    \"I\": \"Ile\",\n",
    "    \"L\": \"Leu\",\n",
    "    \"K\": \"Lys\",\n",
    "    \"M\": \"Met\",\n",
    "    \"F\": \"Phe\",\n",
    "    \"P\": \"Pro\",\n",
    "    \"O\": \"Pyl\",\n",
    "    \"S\": \"Ser\",\n",
    "    \"U\": \"Sec\",\n",
    "    \"T\": \"Thr\",\n",
    "    \"W\": \"Trp\",\n",
    "    \"Y\": \"Tyr\",\n",
    "    \"V\": \"Val\",\n",
    "    \"B\": \"Asx\",\n",
    "    \"Z\": \"Glx\",\n",
    "    \"U\": \"Sec\",\n",
    "    \"X\": \"Xaa\",\n",
    "    \"J\": \"Xle\",\n",
    "    \"*\": \"Stop\",\n",
    "}\n",
    "\n",
    "\n",
    "def annotate_amino_acid_changes(coding_regions, vcf, outfilename):\n",
    "    csvfile = open(vcf, \"r\")\n",
    "    outfile = open(outfilename, \"w\")\n",
    "\n",
    "    reader = csv.reader(csvfile, delimiter=\"\\t\")\n",
    "    writer = csv.writer(outfile, delimiter=\"\\t\")\n",
    "\n",
    "    header = [\n",
    "        \"segment\",\n",
    "        \"gene\",\n",
    "        \"reference_position\",\n",
    "        \"reference_allele\",\n",
    "        \"variant_allele\",\n",
    "        \"coding_region_change\",\n",
    "        \"synonymous/nonsynonymous\",\n",
    "        \"frequency(%)\",\n",
    "        \"frequency\",\n",
    "    ]\n",
    "    writer.writerow(header)\n",
    "\n",
    "    for row in reader:\n",
    "        # ignore comment lines\n",
    "        if \"##\" in row[0] or \"#CHROM\" in row[0]:\n",
    "            continue\n",
    "\n",
    "        segment = row[0]\n",
    "        site = int(row[1]) - 1  # to make this 0 indexed\n",
    "        reference_allele = row[3].lower()\n",
    "        alternative_allele = row[4].lower()\n",
    "        \n",
    "        # pull out the frequency using a string search\n",
    "        SearchStr = r\".+\\:([0-9]{1,2}\\.{0,1}[0-9]{0,2}\\%)\"\n",
    "        result = re.search(SearchStr, row[9])\n",
    "        if result:\n",
    "            frequency = \"\\t\".join(result.groups())\n",
    "            frequency_decimal = frequency.replace(\"%\", \"\")\n",
    "            frequency_decimal = (float(frequency_decimal)) / 100\n",
    "            frequency_decimal = (\n",
    "                \"%.4f\" % frequency_decimal\n",
    "            )  # only include 4 numbers after decimal\n",
    "        else:\n",
    "            frequency = \"none reported\"\n",
    "\n",
    "        for gene in coding_regions[segment]:\n",
    "            gene_name = gene['gene']\n",
    "            cds = gene['coding']\n",
    "            positions = gene['positions']\n",
    "            position = lookup_site(positions, site)\n",
    "            if position is None:\n",
    "                continue\n",
    "            # how far back should we go for the start of the codon?\n",
    "            # zero-based indexing and modular arithmetic pair well here\n",
    "            offset = position % 3\n",
    "            start = position - offset\n",
    "            stop = start + 3\n",
    "            ref_codon = cds[start: stop]\n",
    "            ref_aa = Seq(ref_codon).translate()\n",
    "\n",
    "            # we convert to a list to dynamically modify a base, then join to a string\n",
    "            variant_codon_list = list(ref_codon)\n",
    "            variant_codon_list[offset] = alternative_allele\n",
    "            variant_codon = ''.join(variant_codon_list)\n",
    "            variant_aa = Seq(variant_codon).translate()\n",
    "\n",
    "            # return the amino acid changes, with single letters converted to 3-letter aa codes\n",
    "            if ref_aa == variant_aa:\n",
    "                syn_nonsyn = \"synonymous\"\n",
    "            elif ref_aa != variant_aa and variant_aa == \"*\":\n",
    "                syn_nonsyn = \"stop_gained\"\n",
    "            elif ref_aa != variant_aa:\n",
    "                syn_nonsyn = \"nonsynonymous\"\n",
    "                            \n",
    "            # now determine whether the site is in the 1st, 2nd, or 3rd codon position\n",
    "            # if SNP is in 1st position in codon:\n",
    "            aa_site = int(position / 3) + 1\n",
    "            \n",
    "            amino_acid_change = (\n",
    "                amino_acid_abbreviations[ref_aa]\n",
    "                + str(aa_site)\n",
    "                + amino_acid_abbreviations[variant_aa]\n",
    "            )\n",
    "            output = [\n",
    "                segment,\n",
    "                gene_name,\n",
    "                str(site + 1),\n",
    "                reference_allele.upper(),\n",
    "                alternative_allele.upper(),\n",
    "                amino_acid_change,\n",
    "                syn_nonsyn,\n",
    "                frequency,\n",
    "                frequency_decimal\n",
    "            ]\n",
    "            writer.writerow(output)\n",
    "\n",
    "    csvfile.close()\n",
    "    outfile.close()\n",
    "    return\n",
    "\n",
    "annotate_amino_acid_changes(\n",
    "    coding_regions,\n",
    "    '../data/c2/replicate-1/remapping-2/varscan.vcf',\n",
    "    'varscan-annotated.tsv'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5a562e7-e6c5-4cc7-87a8-b9ed02572b94",
   "metadata": {},
   "source": [
    "This runs! Let's compare this implementation with a previous one that ran. This implementation should be more robust to:\n",
    "- variants that appear in the first exon for a gene with 2+ exons\n",
    "- variants that appear at an intron/exon boundary\n",
    "- frameshift variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5173b11c-e3d0-4633-bb2b-8da68abb87ed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segment</th>\n",
       "      <th>gene</th>\n",
       "      <th>reference_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>synonymous/nonsynonymous</th>\n",
       "      <th>frequency(%)</th>\n",
       "      <th>frequency</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>303</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser92Ser</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.5%</td>\n",
       "      <td>0.0150</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>395</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Glu123Gly</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1.74%</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>498</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Lys157Lys</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.52%</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>546</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Gly173Gly</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>2.4%</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>675</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Arg216Arg</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.19%</td>\n",
       "      <td>0.0419</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>ns</td>\n",
       "      <td>NS1</td>\n",
       "      <td>568</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Leu181Pro</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>2.86%</td>\n",
       "      <td>0.0286</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser57Phe</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>3.01%</td>\n",
       "      <td>0.0301</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>ns</td>\n",
       "      <td>NS1</td>\n",
       "      <td>668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Leu214Leu</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>3.01%</td>\n",
       "      <td>0.0301</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>724</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ile76Val</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1.45%</td>\n",
       "      <td>0.0145</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Leu107Leu</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.27%</td>\n",
       "      <td>0.0127</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    segment gene  reference_position reference_allele variant_allele  \\\n",
       "0       pb2  PB2                 303                T              C   \n",
       "1       pb2  PB2                 395                A              G   \n",
       "2       pb2  PB2                 498                A              G   \n",
       "3       pb2  PB2                 546                A              G   \n",
       "4       pb2  PB2                 675                G              A   \n",
       "..      ...  ...                 ...              ...            ...   \n",
       "181      ns  NS1                 568                T              C   \n",
       "182      ns  NEP                 668                C              T   \n",
       "183      ns  NS1                 668                C              T   \n",
       "184      ns  NEP                 724                A              G   \n",
       "185      ns  NEP                 819                T              C   \n",
       "\n",
       "    coding_region_change synonymous/nonsynonymous frequency(%)  frequency  \\\n",
       "0               Ser92Ser               synonymous         1.5%     0.0150   \n",
       "1              Glu123Gly            nonsynonymous        1.74%     0.0174   \n",
       "2              Lys157Lys               synonymous        1.52%     0.0152   \n",
       "3              Gly173Gly               synonymous         2.4%     0.0240   \n",
       "4              Arg216Arg               synonymous        4.19%     0.0419   \n",
       "..                   ...                      ...          ...        ...   \n",
       "181            Leu181Pro            nonsynonymous        2.86%     0.0286   \n",
       "182             Ser57Phe            nonsynonymous        3.01%     0.0301   \n",
       "183            Leu214Leu               synonymous        3.01%     0.0301   \n",
       "184             Ile76Val            nonsynonymous        1.45%     0.0145   \n",
       "185            Leu107Leu               synonymous        1.27%     0.0127   \n",
       "\n",
       "     Unnamed: 9  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "..          ...  \n",
       "181         NaN  \n",
       "182         NaN  \n",
       "183         NaN  \n",
       "184         NaN  \n",
       "185         NaN  \n",
       "\n",
       "[186 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "pd.read_csv('../data/c13/replicate-2/remapping-2/varscan-annotated.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ecf391de-46cb-4cc2-a4d6-ae573ba19c09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segment</th>\n",
       "      <th>gene</th>\n",
       "      <th>reference_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>synonymous/nonsynonymous</th>\n",
       "      <th>frequency(%)</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>303</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser92Ser</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.5%</td>\n",
       "      <td>0.0150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>395</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Glu123Gly</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1.74%</td>\n",
       "      <td>0.0174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>498</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Lys157Lys</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.52%</td>\n",
       "      <td>0.0152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>546</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Gly173Gly</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>2.4%</td>\n",
       "      <td>0.0240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pb2</td>\n",
       "      <td>PB2</td>\n",
       "      <td>675</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Arg216Arg</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.19%</td>\n",
       "      <td>0.0419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>ns</td>\n",
       "      <td>NS1</td>\n",
       "      <td>568</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Leu181Pro</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>2.86%</td>\n",
       "      <td>0.0286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser57Phe</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>3.01%</td>\n",
       "      <td>0.0301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>ns</td>\n",
       "      <td>NS1</td>\n",
       "      <td>668</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Leu214Leu</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>3.01%</td>\n",
       "      <td>0.0301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>724</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ile76Val</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1.45%</td>\n",
       "      <td>0.0145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>ns</td>\n",
       "      <td>NEP</td>\n",
       "      <td>819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Leu107Leu</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1.27%</td>\n",
       "      <td>0.0127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    segment gene  reference_position reference_allele variant_allele  \\\n",
       "0       pb2  PB2                 303                T              C   \n",
       "1       pb2  PB2                 395                A              G   \n",
       "2       pb2  PB2                 498                A              G   \n",
       "3       pb2  PB2                 546                A              G   \n",
       "4       pb2  PB2                 675                G              A   \n",
       "..      ...  ...                 ...              ...            ...   \n",
       "181      ns  NS1                 568                T              C   \n",
       "182      ns  NEP                 668                C              T   \n",
       "183      ns  NS1                 668                C              T   \n",
       "184      ns  NEP                 724                A              G   \n",
       "185      ns  NEP                 819                T              C   \n",
       "\n",
       "    coding_region_change synonymous/nonsynonymous frequency(%)  frequency  \n",
       "0               Ser92Ser               synonymous         1.5%     0.0150  \n",
       "1              Glu123Gly            nonsynonymous        1.74%     0.0174  \n",
       "2              Lys157Lys               synonymous        1.52%     0.0152  \n",
       "3              Gly173Gly               synonymous         2.4%     0.0240  \n",
       "4              Arg216Arg               synonymous        4.19%     0.0419  \n",
       "..                   ...                      ...          ...        ...  \n",
       "181            Leu181Pro            nonsynonymous        2.86%     0.0286  \n",
       "182             Ser57Phe            nonsynonymous        3.01%     0.0301  \n",
       "183            Leu214Leu               synonymous        3.01%     0.0301  \n",
       "184             Ile76Val            nonsynonymous        1.45%     0.0145  \n",
       "185            Leu107Leu               synonymous        1.27%     0.0127  \n",
       "\n",
       "[186 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotate_amino_acid_changes(\n",
    "    coding_regions,\n",
    "    '../data/c13/replicate-2/remapping-2/varscan.vcf',\n",
    "    'c13-rep2-varscan-annotated.tsv'\n",
    ")\n",
    "new = pd.read_csv('c13-rep2-varscan-annotated.tsv', sep='\\t')\n",
    "new"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc4090f7-fa9f-4da6-822b-0c1cc6c097f0",
   "metadata": {},
   "source": [
    "We observe broad agreement between the tools on the simplest cases. Most notably is that there is agreement on `PA-X` with the new tool up until the slippage site at 594, after which we observe expected disagreement. To further confirm, let's look at amino acids in the translation of PA-X and ensure they compare with the calls. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "04f08ce5-37cf-4eff-9cd7-27501d00bd7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58      Glu2Gly\n",
       "60     Leu42Leu\n",
       "62     Glu59Glu\n",
       "64     Lys73Arg\n",
       "66    Phe105Ser\n",
       "68    Lys134Glu\n",
       "70    Lys142Arg\n",
       "72    Phe148Ser\n",
       "74    Ser193Ser\n",
       "76    Lys202Arg\n",
       "78    Glu230Gly\n",
       "Name: coding_region_change, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pa_x = new.loc[new.gene == 'PA-X', 'coding_region_change']\n",
    "pa_x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d2e342df-1d29-449b-b1a8-da317285e37d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58      2\n",
       "60     42\n",
       "62     59\n",
       "64     73\n",
       "66    105\n",
       "68    134\n",
       "70    142\n",
       "72    148\n",
       "74    193\n",
       "76    202\n",
       "78    230\n",
       "Name: coding_region_change, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positions = pa_x.apply(\n",
    "    lambda x: int(x[3:-3])\n",
    ")\n",
    "positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c27853ed-d579-4322-8c73-5bf2349c7974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58    Glu\n",
       "60    Leu\n",
       "62    Glu\n",
       "64    Lys\n",
       "66    Phe\n",
       "68    Lys\n",
       "70    Lys\n",
       "72    Phe\n",
       "74    Ser\n",
       "76    Lys\n",
       "78    Glu\n",
       "Name: coding_region_change, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref = pa_x.apply(\n",
    "    lambda x: x[:3]\n",
    ")\n",
    "ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "74980a81-96a8-4377-b07d-f82f12cd9732",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Glu', 'Leu', 'Glu', 'Lys', 'Phe', 'Lys', 'Lys', 'Phe', 'Ser', 'Lys', 'Glu']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa = Seq(coding_regions['pa'][1]['coding']).translate()\n",
    "[amino_acid_abbreviations[aa[i-1]] for i in positions]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a671bd03-028e-41b0-9224-3d26debac6eb",
   "metadata": {},
   "source": [
    "This provides a more robust and broadly applicable tool for annotating variants in viral deep sequencing data."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
